rkt/api_service.go

```diff
@@ -0,0 +1,557 @@
+// It returns false if the filter is nil or the pod satisfies the filter, which means
+// it should be returned.
+func filterImage(image *v1alpha.Image, manifest *schema.ImageManifest, filter *v1alpha.ImageFilter) bool {
+	// No filters, return directly.
+	if filter == nil {
+		return false
+	}
+
+	// Filter according to the IDs.
+	if len(filter.Ids) > 0 {
+		if !findString(image.Id, filter.Ids, stringEqual) {
+			return true
+		}
+	}
+
+	// Filter according to the image name prefixes.
+	if len(filter.Prefixes) > 0 {
+		if !findString(image.Name, filter.Prefixes, strings.HasPrefix) {
+			return true
+		}
+	}
+
+	// Filter according to the image base name.
+	if len(filter.BaseNames) > 0 {
+		if !findString(image.Name, filter.BaseNames, hasBaseName) {
+			return true
+		}
+	}
+
+	// Filter according to the image keywords.
+	if len(filter.Keywords) > 0 {
+		if !findString(image.Name, filter.Keywords, strings.Contains) {
+			return true
+		}
+	}
+
+	// Filter according to the imported time.
+	if filter.ImportedAfter > 0 {
+		if image.ImportTimestamp <= filter.ImportedAfter {
+			return true
+		}
+	}
+	if filter.ImportedBefore > 0 {
+		if image.ImportTimestamp >= filter.ImportedBefore {
+			return true
+		}
+	}
+
+	// Filter according to the image labels.
+	if len(filter.Labels) > 0 {
+		if !findKeyValue(manifest.Labels, filter.Labels) {
+			return true
+		}
+	}
+
+	// Filter according to the annotations.
+	if len(filter.Annotations) > 0 {
+		if !findKeyValue(manifest.Annotations, filter.Annotations) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s *v1AlphaAPIServer) ListImages(ctx context.Context, request *v1alpha.ListImagesRequest) (*v1alpha.ListImagesResponse, error) {
+	aciInfos, err := s.store.GetAllACIInfos(nil, false)
+	if err != nil {
+		log.Printf("Failed to get all ACI infos: %v", err)
+		return nil, err
+	}
+
+	var images []*v1alpha.Image
+	for _, aciInfo := range aciInfos {
+		image, manifest, err := s.aciInfoToV1AlphaAPIImage(aciInfo)
+		if err != nil {
+			continue
+		}
+
+		// Filter images.
+		if !filterImage(image, manifest, request.Filter) {
+			images = append(images, image)
+		}
+	}
+
+	return &v1alpha.ListImagesResponse{Images: images}, nil
+}
+
+func (s *v1AlphaAPIServer) InspectImage(ctx context.Context, request *v1alpha.InspectImageRequest) (*v1alpha.InspectImageResponse, error) {
+	aciInfo, err := s.store.GetACIInfoWithBlobKey(request.Id)
+	if err != nil {
+		log.Printf("Failed to get ACI info for image ID %q: %v", request.Id, err)
+		return nil, err
+	}
+
+	image, _, err := s.aciInfoToV1AlphaAPIImage(aciInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1alpha.InspectImageResponse{Image: image}, nil
+}
+
+func (s *v1AlphaAPIServer) GetLogs(request *v1alpha.GetLogsRequest, server v1alpha.PublicAPI_GetLogsServer) error {
+	return fmt.Errorf("not implemented yet")
+}
+
+func (s *v1AlphaAPIServer) ListenEvents(request *v1alpha.ListenEventsRequest, server v1alpha.PublicAPI_ListenEventsServer) error {
```

Comment:
I think it's ok as these docs are already in the interface definition. Copy-pasting them will create a burden on syncing the docs. @krnowak